// fusion-plugin-styletron-react/dist-browser-cjs/browser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _fusionCore = require("fusion-core");

var _styletronReact = require("styletron-react");

var _styletronEngineAtomic = require("styletron-engine-atomic");

var _constants = require("./constants.js");

var _legacyProvider = _interopRequireDefault(require("./legacy-provider.js"));

var _injectDeclarationCompatMixin = require("./inject-declaration-compat-mixin.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

/* eslint-env browser */

/* global module */
const StyletronCompat = (0, _injectDeclarationCompatMixin.injectDeclarationCompatMixin)(_styletronEngineAtomic.Client);
let debugEngine;
let engine;
const plugin = true && (0, _fusionCore.createPlugin)({
  deps: {
    prefix: _constants.AtomicPrefixToken.optional
  },
  middleware: ({
    prefix
  }) => (ctx, next) => {
    if (ctx.element) {
      if (!engine) {
        const config = {
          hydrate: document.getElementsByClassName('_styletron_hydrate_')
        };

        if (prefix !== void 0) {
          config.prefix = prefix;
        }

        engine = new StyletronCompat(config);
      }

      if (process.env.NODE_ENV !== "production" && !debugEngine && typeof Worker !== 'undefined') {
        const worker = new Worker(_constants.workerRoute);
        worker.postMessage({
          id: 'init_wasm',
          url: _constants.wasmRoute
        });
        worker.postMessage({
          id: 'set_render_interval',
          interval: 180
        });

        if (module.hot) {
          module.hot.addStatusHandler(status => {
            if (status === 'dispose') {
              worker.postMessage({
                id: 'invalidate'
              });
            }
          });
        }

        debugEngine = new _styletronReact.DebugEngine(worker);
      }

      ctx.element = _react.default.createElement(_styletronReact.Provider, {
        value: engine,
        debug: debugEngine,
        debugAfterHydration: Boolean(debugEngine)
      }, _react.default.createElement(_legacyProvider.default, {
        value: engine
      }, ctx.element));
    }

    return next();
  }
});
var _default = plugin;
exports.default = _default;
